qcodes/sphinx_extensions/parse_parameter_attr.py

```diff
@@ -0,0 +1,212 @@
+
+import inspect
+from typing import Any, Dict, Optional, Tuple, Type, Union
+
+import parso
+from sphinx.util import logging
+from sphinx.util.inspect import safe_getattr
+
+from qcodes.instrument.base import InstrumentBase
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ParameterProxy:
+    """
+
+    An object that acts as a proxy for documenting containing
+    a repr that can be set from a string.
+
+    """
+
+    def __init__(self, repr_str: str):
+        self._repr = repr_str
+
+    def __repr__(self) -> str:
+        """
+        A repr based on the string set in init.
+        """
+        return self._repr
+
+
+def find_class(
+    node: parso.tree.BaseNode, classname: str
+) -> Tuple[parso.python.tree.Class, ...]:
+    """Find all classes in a given Parso node named ``classname``."""
+    nodes = []
+    for child in node.children:
+        if isinstance(child, parso.python.tree.Class) and child.name.value == classname:
+            nodes.append(child)
+        elif isinstance(child, parso.python.tree.Node):
+            nodes.extend(find_class(child, classname))
+    return tuple(nodes)
+
+
+def find_init_func(
+    node: parso.tree.BaseNode,
+) -> Tuple[parso.python.tree.Function, ...]:
+    """Find all ``__init__`` functions in the supplied Parso node."""
+    nodes = []
+    for child in node.children:
+        if (
+            isinstance(child, parso.python.tree.Function)
+            and child.name.value == "__init__"
+        ):
+            nodes.append(child)
+        elif isinstance(child, parso.python.tree.Node):
+            nodes.extend(find_init_func(child))
+    return tuple(nodes)
+
+
+def parse_init_function_from_str(
+    code: str, classname: str
+) -> Optional[parso.python.tree.Function]:
+    module = parso.parse(code)
+    classes = find_class(module, classname)
+    if len(classes) > 1:
+
+        LOGGER.warning(
+            f"Found more than one class definition for {classname}: Found {classes}"
+        )
+        return None
+    if len(classes) == 0:
+        LOGGER.debug(f"Could not find a class definition for {classname}")
+        return None
+    init_funcs = find_init_func(classes[0])
+    if len(init_funcs) > 1:
+        LOGGER.warning(
+            f"Found more than one init function for {classname}: " f"Found {init_funcs}"
+        )
+        return None
+    if len(init_funcs) == 0:
+        LOGGER.debug(f"Found no init function for {classname}")
+        return None
+    return init_funcs[0]
+
+
+def extract_statements_from_node(
+    parso_node: parso.tree.BaseNode,
+) -> Tuple[parso.python.tree.ExprStmt, ...]:
+    nodes = []
+    for child in parso_node.children:
+        if isinstance(child, parso.python.tree.ExprStmt):
+            nodes.append(child)
+        elif isinstance(child, parso.python.tree.Node):
+            nodes.extend(extract_statements_from_node(child))
+    return tuple(nodes)
+
+
+def eval_params_from_code(code: str, classname: str) -> Dict[str, ParameterProxy]:
```

Comment:
It would probably be worthwhile to cache this using `@functools.lru_cache`  so that each class only needs to be parsed once

Reply:
Done